### ISSUE TITLE
In GraphQL response, "marks" always seems to be an empty array

### ISSUE THREAD

**Issue by tremby** [verbb/vizy#118]
**Description**

When inspecting the query results for a paragraph, `rawNode` shows that I have various "marks" like italic in there, but the `marks` field has just `[]`.

**Steps to reproduce**
1. Make a paragraph with some inlines
2. Query via GraphQL, grabbing at least `rawNode` and `marks` fields
3. Look at response

**Additional info**
- Plugin version: craft-4 7e28546166c2181041f9858db69b0525f6400925
- Craft version: 4.0.x-dev 6a88060
- Multi-site: Yes

**Additional context**
- Headless mode

![screenshot of the issue](https://user-images.githubusercontent.com/199635/162341686-3adb20a2-c35f-4c7f-969e-565f7107d1f3.png)

**Comment by engram-design** [comment]
So this is correct, at least for now. The `marks` property is almost useless as it won't contain anything, pretty much all the time. This is because of how ProseMirror's structure is, and I was starting to add helpers to get you all the marks of a node, but ran into some roadblocks.

To review, here's an example paragraph with some bold and italic text.

```
[
    0 => [
        'type' => 'paragraph'
        'attrs' => [
            'textAlign' => 'start'
        ]
        'content' => [
            0 => [
                'type' => 'text'
                'text' => 'Testing some '
            ]
            1 => [
                'type' => 'text'
                'marks' => [
                    0 => [
                        'type' => 'bold'
                    ]
                ]
                'text' => 'bold formatting'
            ]
            2 => [
                'type' => 'text'
                'text' => ' content with '
            ]
            3 => [
                'type' => 'text'
                'marks' => [
                    0 => [
                        'type' => 'italic'
                    ]
                ]
                'text' => 'italic text as well'
            ]
            4 => [
                'type' => 'text'
                'text' => ', to show marks.'
            ]
        ]
    ]
]
```

Resulting in "Testing some **bold formatting** content with _italic text as well_, to show marks."

My idea was to be able to grab all the marks for a node, but that doesn't really prove all that useful, because what good is getting the bold and italic content, if it's out of order with any other text? So a Text node is the only instance that has marks, and that isn't covered in a GQL fragment.

Open to suggestions if you have any on what you'd expect this behaviour to be, but I'll likely look at removing it as it currently overs nothing of value.

**Comment by tremby** [comment]
I see.

Well, the way both Prismic and the Twitter API do it is to give the entire content of the paragraph, and then alongside that a single array of spans, each of which gives its type, start index, length (or end point), and any extra data needed.
Advantage: it gives support for eg `<em>some running italic with <strong>a bold bit</strong></em>`. (At present Vizy gives a text chunk with `italic`, then another text chunk with both `italic` and `bold` marks. Writing code to not break and restart the `em` between them would be a pain, and important IMO since you don't want a long `a` tag to be broken if part of it has `code` or whatever.)
Disadvantage: consuming it and turning it into HTML or whatever is somewhat tricky (careful with multibyte characters!), and it's possible with bad start and end points to have things like `<strong>this where <em>the tags don't</strong> nest properly</em>`.

It's 6 of one half a dozen of the other as far as I'm concerned.

Given that I couldn't get that information from GQL in a structured way, I went ahead with just asking for `rawNode`, parsing it to an object, and working my way through it. I'm adding some Typescript types as I go, which is helping. It's not that bad at all so far. I'm not sure if I'll do the same for all block types or if I'll want to get the structured fields for some of them. I've only got as far as implementing "paragraph" so far.

So, for now at least, I'd suggest just removing `marks` since it's misleading. I'm not sure of the utility of some of the other fields either.

**Comment by engram-design** [comment]
Appreciate the Prismic resource, that's quite helpful seeing how others are doing it. I really like their approach, and makes sense - however we're largely at the mercy of ProseMirror's schema, using Tiptap.

I'll admit we need to add a nicer layer on top of ProseMirror's data structure to make this a _little_ more developer-friendly, as whole powerful, can be overly verbose and nested - making it difficult to wrangle.

I'll give this some further thought.

**Comment by tremby** [comment]
I tried out my example above...

> Writing code to not break and restart the `em` between them would be a pain, and important IMO since you don't want a long `a` tag to be broken if part of it has `code` or whatever.

1. Make a long link
2. Select part of it and make it italic

This works -- the whole thing is still a link. But...

1. Make a long link
2. Select part of it and make it (inline) code

The link now stops before code starts.

1. Go to HTML mode
2. Make a long link with some code in it like `<a href="/">link with <code>code</code></a>`
3. Switch to wysiwyg mode

The link is broken where the code starts again.

**Comment by tremby** [comment]
And then

1. Make a long italic chunk
2. Make part of it bold
3. See HTML mode -- the italic stops and starts again where bold starts.

But

1. Make a long link
2. Make part of it italic
3. See HTML mode -- a single link tag continues all the way through. (But in GQL it's split into two 😞)

**Comment by engram-design** [comment]
Fixed in [2.1.0](https://github.com/verbb/vizy/releases/tag/2.1.0)
